addons/payment/controllers/portal.py

```diff
@@ -233,12 +231,11 @@ def pay(self, reference='', order_id=None, amount=False, currency_id=None, acqui
             acquirer_domain,
             ['|', ('country_ids', '=', False), ('country_ids', 'in', [partner.country_id.id])]
         ])
+        acquirers = env['payment.acquirer'].search(acquirer_domain)
         if acquirer_id:
-            acquirers = env['payment.acquirer'].browse(int(acquirer_id))
-        if order_id:
-            acquirers = env['payment.acquirer'].search(acquirer_domain)
-        if not acquirers:
-            acquirers = env['payment.acquirer'].search(acquirer_domain)
+            filtered_acq = acquirers.filtered(lambda acq: acq.id == int(acquirer_id))
+            if filtered_acq:
+                acquirers = filtered_acq
```

Comment:
yes, but then you have to check that the acquirer provided in the params:
- is in the correct company
- supports the customer country
- is not disabled

can pbly be done in one sweep with `filtered_domain`, though i'm not sure we'll still have the perf changes you mention here.

Reply:
Ok, so the first part of my comment could be enough:
`acquirers = int(acquirer) in acquirers.ids and acquirers.browse(int(acquirer)) or acquirers`
to avoid to retest your 3 conditions and benefit from the faster comparison in one OP instead of one by pay.acq